### PR TITLE
Add rancher metrics

### PR DESCRIPTION
**Pull request by rmweir**
**Description:**
Added service monitor for rancher. This should provide access to prometheus-client metrics in rancher.

**Issue:**
https://github.com/rancher/rancher/issues/36234

### CODE REVIEW

**Comment by jiaqiluo** [comment]
Although the PR LGTM, I unassigned myself as I am sure about the current requirements for charts. So I will leave it to Arvind and others.

File: packages/rancher-monitoring/rancher-service-metrics/charts/Chart.yaml
Comment (aiyengar2) [inline packages/rancher-monitoring/rancher-service-metrics/charts/Chart.yaml:8]: Scrapes service metrics for Rancher?

File: packages/rancher-monitoring/rancher-service-metrics/charts/templates/servicemonitor.yaml
Comment (aiyengar2) [inline packages/rancher-monitoring/rancher-service-metrics/charts/templates/servicemonitor.yaml:5]: should we error out if we can't find rancher on attempting to install?

File: packages/rancher-monitoring/rancher-service-metrics/charts/templates/servicemonitor.yaml
Comment (aiyengar2) [inline packages/rancher-monitoring/rancher-service-metrics/charts/templates/servicemonitor.yaml:11]: Can we copy the way we do this in the Monitoring chart? e.g. add a helper template function like:

https://github.com/rancher/charts/blob/73f1cd647751650190644b23d0e41221bdfed644/charts/rancher-monitoring/templates/_helpers.tpl#L105-L113

And reference it everywhere like https://github.com/rancher/charts/blob/73f1cd647751650190644b23d0e41221bdfed644/charts/rancher-monitoring/templates/prometheus/prometheus.yaml#L9

Comment (aiyengar2) [reply packages/rancher-monitoring/rancher-service-metrics/charts/templates/servicemonitor.yaml:11]: you can also reuse this in the selector below

**Review by aiyengar2 — CHANGES_REQUESTED** [review]

File: packages/rancher-monitoring/rancher-monitoring/generated-changes/dependencies/rancherServiceMetrics/dependency.yaml
Comment (aiyengar2) [inline packages/rancher-monitoring/rancher-monitoring/generated-changes/dependencies/rancherServiceMetrics/dependency.yaml:2]: can we move this to a different commit and have all the other stuff squashed into the first commit `Add rancher-service-metrics package`?

**Review by aiyengar2 — CHANGES_REQUESTED** [review]

File: packages/rancher-monitoring/rancher-service-metrics/charts/templates/servicemonitor.yaml
Comment (rmweir) [reply packages/rancher-monitoring/rancher-service-metrics/charts/templates/servicemonitor.yaml:5]: Don't think this is necessary, monitoring can be installed without this and if label isn't present the service monitor will just be skipped.

File: packages/rancher-monitoring/rancher-monitoring/generated-changes/dependencies/rancherServiceMetrics/dependency.yaml
Comment (rmweir) [reply packages/rancher-monitoring/rancher-monitoring/generated-changes/dependencies/rancherServiceMetrics/dependency.yaml:2]: Did the first have of this, can you specify which commits to squash together now that it's been updated?

File: charts/rancher-monitoring/100.1.1+up19.0.3/charts/rancherServiceMetrics/templates/servicemonitor.yaml
Comment (cbron) [inline charts/rancher-monitoring/100.1.1+up19.0.3/charts/rancherServiceMetrics/templates/servicemonitor.yaml:24]: Same ? Although would whole use case break if they changed NS ?

File: charts/rancher-monitoring/100.1.1+up19.0.3/charts/rancherServiceMetrics/templates/servicemonitor.yaml
Comment (cbron) [inline charts/rancher-monitoring/100.1.1+up19.0.3/charts/rancherServiceMetrics/templates/servicemonitor.yaml:11]: This name could be more descriptive

File: charts/rancher-monitoring/100.1.1+up19.0.3/charts/rancherServiceMetrics/templates/servicemonitor.yaml
Comment (cbron) [reply charts/rancher-monitoring/100.1.1+up19.0.3/charts/rancherServiceMetrics/templates/servicemonitor.yaml:12]: Oh actually I'm slightly confused, the chart is getting installed into cattle-monitoring-system (from chart.yaml line 4), as well as the service-account, but this ServiceMonitor is going into cattle-system ?

File: charts/rancher-monitoring/100.1.1+up19.0.3/charts/rancherServiceMetrics/templates/servicemonitor.yaml
Comment (rmweir) [reply charts/rancher-monitoring/100.1.1+up19.0.3/charts/rancherServiceMetrics/templates/servicemonitor.yaml:24]: Why would rancher not be installed in cattle-system?

File: charts/rancher-monitoring/100.1.1+up19.0.3/charts/rancherServiceMetrics/templates/servicemonitor.yaml
Comment (rmweir) [reply charts/rancher-monitoring/100.1.1+up19.0.3/charts/rancherServiceMetrics/templates/servicemonitor.yaml:11]: Sure, although it is the service monitor for the rancher service.

File: charts/rancher-monitoring/100.1.1+up19.0.3/charts/rancherServiceMetrics/templates/servicemonitor.yaml
Comment (rmweir) [reply charts/rancher-monitoring/100.1.1+up19.0.3/charts/rancherServiceMetrics/templates/servicemonitor.yaml:53]: @cbron it's the service account used by prometheus to scrape the metrics.

File: charts/rancher-monitoring/100.1.1+up19.0.3/charts/rancherServiceMetrics/templates/servicemonitor.yaml
Comment (cbron) [reply charts/rancher-monitoring/100.1.1+up19.0.3/charts/rancherServiceMetrics/templates/servicemonitor.yaml]: Rancher does have cattle-system hardcoded, so should be fine.

File: charts/rancher-monitoring/100.1.1+up19.0.3/charts/rancherServiceMetrics/templates/servicemonitor.yaml
Comment (cbron) [reply charts/rancher-monitoring/100.1.1+up19.0.3/charts/rancherServiceMetrics/templates/servicemonitor.yaml:11]: Is it for all Rancher things or only for metrics framework ?  `rancher-metrics-service-monitor` ? Maybe Arvind has suggestion.

File: charts/rancher-monitoring/100.1.1+up19.0.3/charts/rancherServiceMetrics/templates/servicemonitor.yaml
Comment (cbron) [reply charts/rancher-monitoring/100.1.1+up19.0.3/charts/rancherServiceMetrics/templates/servicemonitor.yaml:12]: That makes more sense, and cattle-system is hardcoded, ok.

**Review by cbron — COMMENTED** [review]
lgtm, but needs @aiyengar2 approval. ~Also needs to pass CI~.

**Review by aiyengar2 — CHANGES_REQUESTED** [review]
content-wise, this looks good to me.

commit-wise, can you do the following:
- squash commit 2 into commit 1; the new first commit should just be `Add rancher-service-metrics package` with all the changes under `packages/rancher-monitoring/rancher-service-metrics/*` (including the `package.yaml` and the `generated-changes/`)
- commit 3 should only have the change to `packages/rancher-monitoring/rancher-monitoring/generated-changes/dependencies` and `packages/rancher-monitoring/rancher-monitoring/generated-changes/patch` (for the `Chart.yaml.patch` change). Currently, it seems like you committed the prepared chart, which should not be there
- commit 4 should just be a `make charts` commit; there should not be any deletions of `packages/rancher-monitoring/rancher-monitoring/charts` in there.

TLDR:

```bash
git reset HEAD~4
git add packages/rancher-monitoring/rancher-service-metrics/charts
git add packages/rancher-monitoring/rancher-service-metrics/package.yaml
git commit -m "Add rancher-service-metrics package"
git add packages/rancher-monitoring/rancher-monitoring/generated-changes/dependencies/rancherServiceMetrics/dependency.yaml
git add packages/rancher-monitoring/rancher-monitoring/generated-changes/patch/Chart.yaml.patch
git commit -m "Add dependency on rancher-service-metrics"
git add charts
git add assets
git add index.yaml
git commit -m "make charts"
```

**Review by aiyengar2 — APPROVED** [review]
LGTM

**Pull request merged**
